rfcs/53-generalized-hierarchy.md

```diff
@@ -0,0 +1,273 @@
+        too.
+      - A HierarchyEvent is added to the queue to reflect this update.
+  - `move_to_first(Entity)` - Moves one of the children to the front.
+  - `move_to_last(Entity)` - Moves one of the children to the end.
+  - `remove_all_children()` - Disconnects all children from a parent. All
+    children become roots.
+      - A HierarchyEvent is added to the queue to reflect this update.
+
+All of these commands will also update the four marker components appropriately.
+Optionally, we could also mark these components as changed (even though they're
+ZSTs) to leverage change detection.
+
+One optional extension to this design is to create dedicated `HierarchyQuery<Q,
+F>`, which is a custom SystemParam that wraps multiple Query objects to simplify
+querying for items in the hierarchy:
+
+ - `iter_descendants(_mut)` - returns an iterator over `(parent, child)` results
+   from a query parameter. This does a depth-first traversal over the hierarchy
+   starting at some entity.
+ - `iter_ancestors(_mut)` - returns an iterator over `(parent, child)` results
+   from a query parameter. This bubbles up queries from the bottom to the top
+   starting at some entity.
+
+As these operations are handled at any command buffer flush. The hierarchy
+maintenance system can be removed.
+
+## Benefits
+The biggest benefit to this new design is guarenteed global consistency. The
+hierarchy cannot be arbitrarily mutated between synchronization points.
+
+The second benefit is that any update to the hierarchy is "atomic" and is `O(1)`
+for the majority of the operations handled.
+
+## Drawbacks
+There are a notable number of drawbacks to this design:
+
+ - Counting children is now `O(n)`. Finding and editing children order is still
+   `O(n)`, that has not changed.
+ - The fact that the hierarchy is not optimally stored in memory is still an
+   issue, and all hiearchy traversals require heavy random access into memory.
+ - Updates are still not immediately visible and deferred until the next command
+   buffer flushes.
+ - Hierarchy updates are now single threaded.
+ - The marker components may cause heavy archetype fragmentation.
+ - Some potential implementations of this hierarchy requires normal ECS commands
+   to be aware of the hierarchy. For example, despawning a child in the middle of
+   the linked list without updating it's immediate siblings will violate several
+   invariants of the system.
+
+## Rationale and alternatives
+The primary goal of this proposal is to get rid of the hierarchy maintenance
+system and the consistency issues that come with it's design, which it delivers
+on.
+
+### Keep `Parent`/`Children` as is. Implement Commands/Events.
+This simply keeps the components as is but restricts both of them to be
+read-only. The commands and events seen in this proposal can also be implemented
+on top of them. The internals of `Children` could be replaced with a `HashSet<T>`
```

Comment:
I think that replacing this with a `HashSet` is likely to be dramatically slower in practice: asymptotics are irrelevant for tiny n.

Reply:
Part of this I think is also to ensure child uniqueness, which has been noted as an issue in the past. We can't just `push` a child on to a Vec and call it a day here. Though I do think the most egregious child counts will be seen in UI use cases (i.e. a list of items in a scroll view could easily number up several hundred if not properly controlled). Though even then the entire children array fits in cache, so it may not be that bad of a check.

Reply:
Sure, but we can use other forms of sets that have better small-n performance :)